core/node/node_framework/src/service/named_future.rs

```diff
@@ -0,0 +1,51 @@
+use std::{fmt, future::Future, pin::Pin, task};
+
+use pin_project_lite::pin_project;
+
+use crate::task::TaskId;
+
+pin_project! {
+    /// Implements a future with the name tag attached.
+    pub struct NamedFuture<F> {
+        #[pin]
+        inner: F,
+        name: TaskId,
+    }
+}
+
+impl<T, F> NamedFuture<F>
```

Comment:
Dumb question: Why do you need `T` param here? It looks like it's possible to define `F: Future` w/o specifying its output type. Ditto below (you can use `type Output = F::Output`).